File: leetcode-110_balanced-binary-tree.go
/**
 * 给定一个二叉树, 判断是否平衡二叉树
 * 平衡二叉树：空树, 或左右子树的高度差不超过1, 并且左右子树都是平衡二叉树
 */
func depth(root *TreeNode) int {
    if root == nil {
        return 0
    }
    return int(math.Max(float64(depth(root.Left)), float64(depth(root.Right)))) + 1
}

func isBalanced(root *TreeNode) bool {
    //空树
    if root == nil {
        return true
    }
    leftDepth := depth(root.Left)
    rightDepth := depth(root.Right)
    //左右子树的高度差不超过1, 并且左右子树都是平衡二叉树
    return math.Abs(float64(leftDepth - rightDepth)) <= 1 && isBalanced(root.Left) && isBalanced(root.Right)
}

File: leetcode-110_balanced-binary-tree.java
/**
 * 判断二叉树是否平衡二叉树
 */
class Solution {
    private boolean balance = true;

    // 深度优先遍历，自底向上判断子树是否平衡，避免自顶向下递归的重复计算高度
    public int depth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int leftDepth = depth(root.left);
        int rightDepth = depth(root.right);
        if (Math.abs((leftDepth - rightDepth)) > 1) {
            balance = false;
        }
        return Math.max(leftDepth, rightDepth) + 1;
    }
    public boolean isBalanced(TreeNode root) {
        depth(root);
        return balance;
    }
}
